Manifold/Expression.swift

```diff
@@ -333,67 +338,72 @@ extension Expression where Recur: FixpointType, Recur: Equatable {
 				.flatMap { A in
 					A.analysis(
 						ifLambda: { i, type, body in
-							b.typecheck(environment, against: type.out).map { body.out.substitute(i, $0) }
+							b.typecheck(context, against: type.out).map { body.out.substitute(i, $0) }
 						},
 						otherwise: const(Either.Left("illegal application of \(a) : \(A) to \(b)")))
 				}
 
 		case let .Projection(term, branch):
-			return term.typecheck(environment)
+			return term.typecheck(context)
 				.flatMap { type in
 					type.analysis(
 						ifLambda: { i, A, B in
 							Either.Right(branch ? B.out.substitute(i, A.out) : A.out)
 						},
-						otherwise: const(Either.Left("illegal attempt to project field \(branch ? 1 : 0) of value \(term) of non-product type \(type)")))
-				}
-
-		// Typecheck products annotated with lambda type as sums.
-		case let .Annotation(.Product(tag, body), .Lambda(i, tagType, typeBody)):
-			return tagType.out.typecheck(environment, against: .Type(0))
-				.flatMap { _ in
-					tag.out.typecheck(environment, against: tagType.out)
-						.flatMap { _ in
-							typeBody.out.substitute(i, tag.out).typecheck(environment)
-								.flatMap { bodyType in
-									body.out.typecheck(environment, against: bodyType)
-										.map(const(.Lambda(i, tagType, typeBody)))
-								}
-						}
+						otherwise: const(Either.Left("illegal projection of field \(branch ? 1 : 0) of non-product value \(term) of type \(type)")))
 				}
 
 		case let .Annotation(term, type):
-			return type.typecheck(environment, against: .Type(0))
-				.map { $0.evaluate() }
-				.flatMap { type in term.typecheck(environment, against: type) }
+			return term.typecheck(context, against: type)
+				.map(const(type))
 		}
 	}
 
-	public func typecheck(environment: [Name: Expression], against: Expression) -> Either<Error, Expression> {
-		return typecheck(environment)
-			.flatMap { type in
-				type == against || against == .Type(0) && type.isType
-					? Either.Right(type)
-					: Either.Left("Type mismatch: expected \(String(reflecting: self)) to be of type \(String(reflecting: against)), but it was actually of type \(String(reflecting: type)) in environment \(environment)")
-		}
+	public func typecheck(context: [Name: Expression], against: Expression) -> Either<Error, Expression> {
+		return (against.isType
+				? Either.Right(against)
+				: against.typecheck(context, against: .Type(0)))
+			.map { _ in against.evaluate() }
+			.flatMap { against in
+				typecheck(context)
+					.map { $0.evaluate() }
+					.flatMap { (type: Expression) -> Either<Error, Expression> in
+						if case let (.Product(tag, payload), .Lambda(i, tagType, body)) = (self, against) {
+							return tagType.out.typecheck(context, against: .Type(0))
+								.flatMap { _ in
+									tag.out.typecheck(context, against: tagType.out)
+										.flatMap { _ in
+											payload.out.typecheck(context, against: body.out.substitute(i, tag.out))
+												.map(const(type))
+										}
+								}
+						}
```

Comment:
By moving annotated product checking here, we can consolidate our handling of certain niceties such as evaluating the type expressions. Where it was, an annotation of a product type such as `(false, ()) : Natural` wouldn’t be caught by the check, because `Natural` is a named variable in the environment/context, and therefore _evaluates_ to a lambda, rather than already being one.

This amounts to the premises in our derivation tree; `typecheck(_:against:)` is predicated on `against` being a type, and in normal form.